File: 2-second-year-2022-2023/Programming-languages-and-laboratory/Java/Exercises/19_compito_java_2018-01-29/GestioneAcquedotto.java
import java.io.*;
import java.util.*;

public class GestioneAcquedotto {

	public static void main(String[] args) {
		List<Cliente> clienti = new LinkedList<Cliente>();
		Map<Integer, Cliente> id_to_cl = new HashMap<Integer, Cliente>();
		double max = 0;
		String max_cl = "";

		// PUNTO 1: memorizzazione clienti
		try {
			BufferedReader br = new BufferedReader(new FileReader("clienti.txt"));
			String line = br.readLine(); // prima riga: tipo+codice
			while (line != null) {
				// lettura di 1 cliente
				String nome;
				String indirizzo;
				String ragioneSociale;
				String tipo_contratto;
				double portata;
				int codice;
				StringTokenizer tokenizer = new StringTokenizer(line);
				String type = tokenizer.nextToken();
				int id = Integer.parseInt(tokenizer.nextToken());
				if (type.equals("residenziale")) {
					nome = br.readLine();
					tipo_contratto = br.readLine();
					indirizzo = br.readLine();
					Cliente c = new Residenziale(id, indirizzo, tipo_contratto.equals("canone concordato"), nome);
					clienti.add(c);
					id_to_cl.put(id, c);

				} else {
					ragioneSociale = br.readLine();
					portata = Double.parseDouble(br.readLine());

					tipo_contratto = br.readLine();

					indirizzo = br.readLine();
					Cliente c = new Aziendale(id, indirizzo, tipo_contratto.equals("canone concordato"), ragioneSociale,
							portata);
					clienti.add(c);
					id_to_cl.put(id, c);
				}
				line = br.readLine();
				line = br.readLine();
			} // while
			br.close();
		} catch (IOException e) {
			System.err.println(e);
		} catch (Exception e) {
			System.err.println(e);
		}
		// PUNTO 2
		try {
			BufferedReader br = new BufferedReader(new FileReader("letture.txt"));
			String line = br.readLine();
			while (line != null) {
				StringTokenizer tokenizer = new StringTokenizer(line);
				int id = Integer.parseInt(tokenizer.nextToken());
				double vol = Double.parseDouble(tokenizer.nextToken());
				Cliente c = id_to_cl.get(id);
				c.aggiungiLettura(vol);
				if (vol > max) {
					max_cl = c.getNome();
					max = vol;
				}
				line = br.readLine();
			} // while
			br.close();
		} catch (IOException e) {
			System.err.println(e);
		} catch (Exception e) {
			System.err.println(e);
		}
		// punto 3
		System.out.println("ID\tTipo\tNome/Cognome\tRagioneSociale\tIndirizzo\tTipoContratto\tPortataMassima");
		for (Cliente c : clienti)
			System.out.println(c);

		System.out.println();

		// punto 4
		System.out.println("ID\tNome/Cognome/RagioneSociale\tTotale");
		for (Cliente c : clienti) {
			double bolletta;
			if (c.getConcordato())
				bolletta = c.getTotaleLet() * 1.15;
			else
				bolletta = c.getTotaleLet() * 1.2;

			System.out.println(c.getId() + "\t" + c.getNome() + "\t" + bolletta);
		}

		// PUNTO 5
		System.out.println("\nCliente con bolletta piu' alta: " + max_cl + "\t" + max);
	}// main
}// class

File: 2-second-year-2022-2023/Programming-languages-and-laboratory/Python/Exercises/19_compito_java_20180129/GestioneAcquedotto.py
from Cliente import Cliente
from Residenziale import Residenziale
from Aziendale import Aziendale
from Lettura import Lettura


class GestioneAcquedotto:
    def main():

        clienti = []
        cod_to_cliente = {}
        letture = []
        
        try:
            f = open("clienti.txt", "r")
            line = f.readline().strip()

            while line != "":
                tok = line.split()
                tipo = tok[0]
                codice = int(tok[1])
                nome_cognome_ragione_sociale = f.readline().strip()

                if tipo == "residenziale":
                    tipo_contratto = f.readline().strip()
                    indirizzo = f.readline().strip()

                    res = Residenziale(
                        tipo, codice, nome_cognome_ragione_sociale, tipo_contratto, indirizzo)
                    clienti.append(res)
                    cod_to_cliente[codice] = res

                    line = f.readline().strip()

                else:
                    portata_massima = float(f.readline().strip())
                    tipo_contratto = f.readline().strip()
                    indirizzo = f.readline().strip()

                    az = Aziendale(tipo, codice, nome_cognome_ragione_sociale,
                                   portata_massima, tipo_contratto, indirizzo)
                    clienti.append(az)
                    cod_to_cliente[codice] = az

                    line = f.readline().strip()

                line = f.readline().strip()

        except IOError as i:
            print(i)
        except Exception as e:
            print(e)

        print("ID Tipo Nome/Cognome RagioneSociale Indirizzo TipoContratto PortataMassima")

        for c in clienti:
            print(c)
        print()

        try:
            f = open("letture.txt", "r")
            line = f.readline().strip()

            while line != "":
                tok = line.split()
                codice_cliente = int(tok[0])
                metri_cubi_consumati = float(tok[1])
                cod_to_cliente[codice_cliente].add_metri_cubi(
                    metri_cubi_consumati)
                let = Lettura(codice_cliente, metri_cubi_consumati)
                letture.append(let)
                line = f.readline().strip()

        except IOError as i:
            print(i)
        except Exception as e:
            print(e)

        print("ID Nome/Cognome/RagioneSociale Totale")

        for cl in clienti:
            if cl.get_tipo_contratto() == "mercato libero":
                totale = cl.get_metri_cubi() * 1.2
            else:
                totale = cl.get_metri_cubi() * 1.15
            print(str(cl.get_codice()) + " " +
                  cl.get_nome() + " " + str(totale))
        print()
            
        lettura_max = 0.0
        nome_cliente_lettura_max = ""
        for let in letture:
            if let.get_lettura() > lettura_max:
                lettura_max = let.get_lettura()
                nome_cliente_lettura_max = cod_to_cliente[let.get_codice_cliente()].get_nome()
                
        print("Il cliente con la lettura più alta è " + nome_cliente_lettura_max + " con " + str(lettura_max))


if __name__ == "__main__":
    GestioneAcquedotto.main()
